### PR TITLE
Pipeline logging

### PR DESCRIPTION
**Pull request by nicHoch**

### CODE REVIEW

File: stixcore/processing/pipeline.py [hunk 1]
```diff
@@ -50,36 +58,163 @@ def __init__(self, func, regex):
         if not isinstance(regex, type(re.compile('.'))):
             raise TypeError('regex must be a regex Pattern')
         self.regex = regex
+        self.args = args
+        self.queue = Queue(maxsize=0)
+        self.name = name
+        self.lp = threading.Thread(target=self.process)
+        self.lp.start()
+
+    @poll_decorator(step=1, poll_forever=True)
```
Comment (samaloney) [inline stixcore/processing/pipeline.py:67]: So you poll but only will the `queue.get` not wait forever at each poll?

Comment (nicHoch) [reply stixcore/processing/pipeline.py:67]: yes the  queue.get will wait until something new comes in.
But after that the process is done.
the poll_decorator starts a new process afterwards again and again .

File: stixcore/processing/pipeline.py [hunk 2]
```diff
@@ -53,35 +59,145 @@ def __init__(self, func, regex, **args):
...
+        logging.getLogger().removeHandler(self.fh)
+        if not self.allright:
+            shutil.copyfile(self.log_file, self.err_file)
+            if CONFIG.getboolean('Pipeline', 'error_mail_send', fallback=False):
+                try:
+                    sender = CONFIG.get('Pipeline', 'error_mail_sender', fallback='localhost')
+                    receivers = CONFIG.get('Pipeline', 'error_mail_receivers').split(",")
+                    host = CONFIG.get('Pipeline', 'error_mail_smpt_host', fallback='localhost')
+                    port = CONFIG.getint('Pipeline', 'error_mail_smpt_port', fallback=25)
+                    smtpObj = smtplib.SMTP(host=host, port=port)
```
Comment (samaloney) [inline stixcore/processing/pipeline.py:139]:
```suggestion
                    smtp_server = smtplib.SMTP(host=host, port=port)
```

File: stixcore/processing/pipeline.py [hunk 3]
```diff
@@ -53,35 +59,145 @@ def __init__(self, func, regex, **args):
...
+
+{self.err_file}
+
+StixCore
+
+==========================
+
+do not answer to this mail.
+"""
+                    smtpObj.sendmail(sender, receivers, message)
```
Comment (samaloney) [inline stixcore/processing/pipeline.py:154]: For test you could mock this method and just make it has the right contents. Now if you really want to test if the email is sent that won't work but I think that would be an external issue to stixcore

File: stixcore/processing/tests/test_processing.py [hunk 4]
```diff
@@ -268,11 +270,69 @@ def test_single_vs_batch(out_dir):
...
+    finally:
+        CONFIG.set('Logging', 'stop_on_error', str(CONTINUE_ON_ERROR))
+        CONFIG.set('Paths', 'fits_archive', str(FITS_ARCHIVE))
+        CONFIG.set('Pipeline', 'log_level', str(LOG_LEVEL))
+        CONFIG.set('Pipeline', 'log_dir', str(LOG_DIR))
 
-    currently used to start the AUX processing on an existing L1 fits dir
-    '''
+
+@pytest.mark.skipif(not CONFIG.getboolean('Pipeline', 'error_mail_send', fallback=False),
```
Comment (samaloney) [inline stixcore/processing/tests/test_processing.py:304]: As mentioned above you could mock this so the test could run

File: tox.ini [hunk 5]
```diff
@@ -42,7 +42,7 @@ deps =
 extras =
     test
 commands =
-    pytest --pyargs stixcore {toxinidir}/docs --cov stixcore --cov-config={toxinidir}/setup.cfg --remote-data=any -m "not end2end" {posargs}
+    pytest --pyargs stixcore {toxinidir}/docs --cov stixcore --cov-config={toxinidir}/setup.cfg --remote-data=any -m "not end2end" -n 1 --dist loadfile {posargs}
```
Comment (samaloney) [inline tox.ini:45]: Is the distributed test part of this PR?

File: stixcore/idb/tests/test_idb.py [hunk 6]
```diff
@@ -17,10 +17,12 @@
 
 
 @pytest.fixture
+@pytest.mark.remote_data
```
Comment (samaloney) [inline stixcore/idb/tests/test_idb.py:20]: maybe in the test config we should download the IDB so this isn't necessary? seem awkward

File: setup.cfg [hunk 7]
```diff
@@ -23,6 +23,7 @@ install_requires =
     pyyaml
     intervaltree
     hissw
+    polling2
```
Comment (samaloney) [inline setup.cfg:26]: I guess I'm a little bit concerned about having a dependancy on a one man library

File: stixcore/processing/pipeline.py (hunk 1 above)
Comment (samaloney) [reply stixcore/processing/pipeline.py:67]: What the advantage compare to doing something like this inside the function? 
```
while True:
    path = self.queue.get()
    try:
         self.func(path, **self.args)
    except Exception as e:
        logger.error(e)
             if CONFIG.getboolean('Logging', 'stop_on_error', fallback=False):
                 raise e

**Review by samaloney — APPROVED** [review]

File: tox.ini (hunk 5 above)
Comment (nicHoch) [reply tox.ini:45]: the preparations somehow yes 
but n is set to 1 so ineffective at the moment

File: setup.cfg (hunk 7 above)
Comment (nicHoch) [reply setup.cfg:26]: yes but this is a very isolated function and location in the code
if problems show up we can replace it

File: stixcore/processing/tests/test_processing.py (hunk 4 above)
Comment (nicHoch) [reply stixcore/processing/tests/test_processing.py]: good point will do so

**Pull request merged**